stan/math/prim/scal/fun/promoter.hpp

```diff
@@ -0,0 +1,41 @@
+#ifndef STAN_MATH_PRIM_SCAL_FUN_PROMOTER_HPP
+#define STAN_MATH_PRIM_SCAL_FUN_PROMOTER_HPP
+
+namespace stan {
+  namespace math {
+    /**
+     * Struct which holds static function for element-wise type promotion.
+     *
+     * @tparam F type of input element
+     * @tparam T type of output element
+     */
+    template <typename F, typename T>
+    struct promoter {
+      inline static void promote(const F& u, T& t) {
+        t = u;
+      }
+      inline static T promote_to(const F& u) {
+        return u;
+      }
+    };
+
+    /**
+     * Struct which holds static function for element-wise type promotion.
+     * This specialization is for scalars of the same type.
+     *
```

Comment:
I don't think you changed this, but why do we need a specialization for two things of the same type? I see this pattern in a couple of places and it seems to have an identical implementation to the one for two different types, so I'm confused about what is going on here / why we need this in general...